lms/static/scripts/frontend_apps/components/YouTubePicker.tsx

```diff
@@ -43,6 +49,18 @@ export default function YouTubePicker({
     }
   };
 
+  // Display an error if a videoId was resolved (meaning, the URL fulfils the right pattern), but the server returns
+  // an error
+  useEffect(() => {
+    if (videoInfo.error) {
+      transitionToErrorState(
+        videoInfo.error === 'video_not_found'
+          ? 'Provided URL does not belong to a valid YouTube video'
+          : undefined
+      );
+    }
+  }, [videoInfo]);
```

Comment:
Thanks!

This definitely felt wrong when I was writing it. It's something I wanted to revisit.

Thanks for the suggestions and the documentation link. I'll try to put something together with that.

Reply:
Ok, so I have come up with something, but it's still not perfect. Let me follow my track of thought for context:

First I have tried to derive and memoize `error` from other states, assuming the possibilities are:

* There's a fetch error: That means the video URL matches the pattern, but the API failed. Probably the video is not found, but we fallback to another error.
* There's no fetch error, and the videoId is null: The URL is invalid
* In any other case, there's no error.

```ts
const [videoId, setVideoId] = useState(
  (defaultURL && videoIdFromYouTubeURL(defaultURL)) || null
);
const videoInfo = useAPIFetch<YouTubeVideoInfo>(
  videoId ? urlPath`/api/youtube/videos/${videoId}` : null
);
const error = useMemo(() => {
  const defaultError =
    'URL must be a YouTube video, e.g. "https://www.youtube.com/watch?v=cKxqzvzlnKU"';

  // We have a fetch error. That means a valid URL was set, but the API could not find the video
  if (videoInfo.error && isAPIError(videoInfo.error)) {
    return videoInfo.error.errorCode === 'youtube_video_not_found'
      ? 'Video not found'
      : defaultError;
  }

  // The videoId is not valid. That means a URL was set but it's not a valid YouTube URL
  if (!videoId) {
    return defaultError;
  }

  return undefined;
}, [videoId, videoInfo.error]);
```

The problem with this approach is that the initial value of `videoId` is null, so an error is displayed right from the beginning.

My next attempt has been trying to track the last set URL as you suggested, to be able to tell between "no URL is set yet" and "a URL was set but it's not valid".

```ts
const [videoId, setVideoId] = useState(
  (defaultURL && videoIdFromYouTubeURL(defaultURL)) || null
);
const lastVideoURL = useRef<string>();
const videoInfo = useAPIFetch<YouTubeVideoInfo>(
  videoId ? urlPath`/api/youtube/videos/${videoId}` : null
);
const error = useMemo(() => {
  const defaultError =
    'URL must be a YouTube video, e.g. "https://www.youtube.com/watch?v=cKxqzvzlnKU"';

  // We have a fetch error. That means a valid URL was set, but the API could not find the video
  if (videoInfo.error && isAPIError(videoInfo.error)) {
    return videoInfo.error.errorCode === 'youtube_video_not_found'
      ? 'Video not found'
      : defaultError;
  }

  // The videoId is not sset. That means a URL was set but it's not a valid YouTube URL
  if (!videoId && lastVideoURL.current) {
    return defaultError;
  }

  return undefined;
}, [videoId, videoInfo.error]);

// [...]

const onURLChange = (inputURL: string) => {
  lastVideoURL.current = inputURL;
  setVideoId(videoIdFromYouTubeURL(inputURL));
};
```

The problem with this is that it doesn't seem to be possible to use a ref as a `useMemo` dependency, which means the error is not re-computed when transitioning from "no URL is set yet" to "a URL was set but it's not valid", because `videoId` is still null, so "nothing" changed in practice.

My last attempt has been tracking all the states directly in `videoId`, making it `string | null | 'INVALID'`.

* `null`: No URL is set.
* `string`: a matching YouTube URL was set.
* `'INVALID'`: a non-matching/invalid URL was set.

```ts
const [videoId, setVideoId] = useState<string | null | 'INVALID'>(
  (defaultURL && videoIdFromYouTubeURL(defaultURL)) || null
);
const videoInfo = useAPIFetch<YouTubeVideoInfo>(
  videoId && videoId !== 'INVALID'
    ? urlPath`/api/youtube/videos/${videoId}`
    : null
);
const error = useMemo(() => {
  const defaultError =
    'URL must be a YouTube video, e.g. "https://www.youtube.com/watch?v=cKxqzvzlnKU"';

  // We have a fetch error. That means a valid URL was set, but the API could not find the video
  if (videoInfo.error && isAPIError(videoInfo.error)) {
    return videoInfo.error.errorCode === 'youtube_video_not_found'
      ? 'Video not found'
      : defaultError;
  }

  // The videoId is not valid. That means a URL was set but it's not a valid YouTube URL
  if (videoId === 'INVALID') {
    return defaultError;
  }

  return undefined;
}, [videoId, videoInfo.error]);

// [...]

const onURLChange = (inputURL: string) => {
  setVideoId(videoIdFromYouTubeURL(inputURL) ?? 'INVALID');
};
```

This one works, but I'm not sure if it's clear enough.

For comparison, in JSTORPicker we use a different approach, which involves a helper error state, and then there's an on-the-fly computing of `renderedError` based on that state, and whatever comes from fetch calls, but with no memoization.

I'm now trying to do something similar for YouTubePicker, but I don't love that approach either.

Reply:
I have found an approach that I don't hate, which is the second option above, but instead of tracking `videoId` as `string | null | 'INVALID'`, I track `currentVideo` with type `{ isInvalid: false; videoId: string | null } | { isInvalid: true }`.

That allows using `isInvalid` as type guard, but the object can have three possible shapes:

* `{ isInvalid: false; videoId: null }`: Initial state. No URL is set yet.
* `{ isInvalid: false; videoId: '...' }`: A URL matching YouTube is set. We can check fetch error separately.
* `{ isInvalid: true }`: A URL is set, but it does not match YouTube pattern.

```ts
type CurrentVideo =
  | {
      isInvalid: false;
      videoId: string | null;
    }
  | {
      isInvalid: true;
    };

const [currentVideo, setCurrentVideo] = useState<CurrentVideo>({
  isInvalid: false,
  videoId: (defaultURL && videoIdFromYouTubeURL(defaultURL)) || null,
});
const videoInfo = useAPIFetch<YouTubeVideoInfo>(
  !currentVideo.isInvalid && currentVideo.videoId
    ? urlPath`/api/youtube/videos/${currentVideo.videoId}`
    : null
);
const error = useMemo(() => {
  const defaultError =
    'URL must be a YouTube video, e.g. "https://www.youtube.com/watch?v=cKxqzvzlnKU"';

  // We have a fetch error. A valid URL was set, but the API could not find the video
  if (videoInfo.error && isAPIError(videoInfo.error)) {
    return videoInfo.error.errorCode === 'youtube_video_not_found'
      ? 'Video not found'
      : defaultError;
  }

  // A URL was set but it's not a valid YouTube URL
  if (currentVideo.isInvalid) {
    return defaultError;
  }

  return undefined;
}, [currentVideo, videoInfo.error]);

const onURLChange = (inputURL: string) => {
  const videoId = videoIdFromYouTubeURL(inputURL);
  setCurrentVideo(
    videoId ? { videoId, isInvalid: false } : { isInvalid: true }
  );
};
```

Reply:
And an even better approach. Thanks @robertknight for the suggestion:

```ts
const [currentURL, setCurrentURL] = useState(defaultURL);
const videoId = useMemo(
  () => (currentURL ? videoIdFromYouTubeURL(currentURL) : null),
  [currentURL]
);
const videoInfo = useAPIFetch<YouTubeVideoInfo>(
  videoId ? urlPath`/api/youtube/videos/${videoId}` : null
);
const error = useMemo(() => {
  const defaultError =
    'URL must be a YouTube video, e.g. "https://www.youtube.com/watch?v=cKxqzvzlnKU"';

  // We have a fetch error. A valid YouTube URL was set, but the API could not find the video
  if (videoInfo.error && isAPIError(videoInfo.error)) {
    return videoInfo.error.errorCode === 'youtube_video_not_found'
      ? 'Video not found'
      : defaultError;
  }

  // A URL was set, but it's not a valid YouTube URL
  if (currentURL && !videoId) {
    return defaultError;
  }

  return undefined;
}, [currentURL, videoId, videoInfo.error]);

const onURLChange = (inputURL: string) => setCurrentURL(inputURL);
```